include/xtensor/xrandom.hpp

```diff
@@ -974,6 +986,96 @@ namespace xt
+         *
+         * Note: this function makes a copy of your data, and only 1D data is accepted.
+         *
+         * @param e expression to sample from
+         * @param n number of elements to sample
+         * @param e expression for the weight distribution.
+         *          Weights must be positive and real-valued but need not sum to 1.
+         * @param replace whether to sample with or without replacement
+         * @param engine random number engine
+         *
+         * @return xtensor containing 1D container of sampled elements
+         */
+        template <class T, class W, class E>
+        xtensor<typename T::value_type, 1>
+        choice(const xexpression<T>& e, std::size_t n, const xexpression<W>& weights, bool replace, E& engine)
+        {
+            const auto& de = e.derived_cast();
+            const auto& dweights = weights.derived_cast();
+            if (de.dimension() != 1)
+            {
+                XTENSOR_THROW(std::runtime_error, "Sample and weight expression must be 1 dimensional");
+            }
+            if (de.size() < n && !replace)
+            {
+                XTENSOR_THROW(std::runtime_error, "If replace is false, then the sample expression's size must be > n");
+            }
+            if (de.size() != dweights.size() || de.dimension() != dweights.dimension())
+            {
+                XTENSOR_THROW(std::runtime_error, "Sample and weight expression must have the same size");
+            }
+            XTENSOR_ASSERT(std::all_of(dweights.begin(), dweights.end(), [](auto w){ return w >= 0; }));
+            static_assert(std::is_floating_point<typename W::value_type>::value,
+                          "Weight expression must be of floating point type");
+            using result_type = xtensor<typename T::value_type, 1>;
+            using size_type = typename result_type::size_type;
+            using weight_type = typename W::value_type;
+            result_type result;
+            result.resize({n});
+
+            if (replace)
+            {
+                // Sample u uniformly in the range [0, sum(weights)[
+                // The index idx of the sampled element in e is the largest idx such that weight_cumul[idx] < u (given by std::upper_bound - 1).
+                const xtensor<weight_type, 1> weight_cumul = cumsum(dweights);  // 0 included as first elem
+                const auto weight_cumul_begin = weight_cumul.storage().begin();
+                std::uniform_real_distribution<weight_type> weight_dist{0, weight_cumul[weight_cumul.size() - 1]};
+                for(auto& x : result)
+                {
+                    const auto u = weight_dist(engine);
+                    const auto idx_iter = std::upper_bound(weight_cumul_begin, weight_cumul.storage().end(), u) - 1;
+                    const auto idx = static_cast<size_type>(idx_iter - weight_cumul_begin);
+                    x = de.storage()[idx];
```

Comment:
The data needs to be one dimensional.

This is also how it was done in the other overload, but I am not sure when to access data with `.data()`, `.storage()`, or directly with `operator()`.

Reply:
So why does the data need to be 1d?

Reply:
The element in `e` sampled from are meant to represent a set. If the data was n-dimensional, I don't know what the output would be. Without replacement it could be a masked array, but with replacement it is not clear.

Another possibility would be that if the input is n-dimensional, then iwe do internally `e.reshape(-1)`.
In the Numpy do, we have
> a : 1-D array-like or int
    If an ndarray, a random sample is generated from its elements.
    If an int, the random sample is generated as if a were np.arange(a)

Reply:
It seems sensible to effectively treat it as 1d both allow nd input. But maybe I'm overlooking something.

Reply:
> but I am not sure when to access data with .data(), .storage(), or directly with operator().

`storage` returns a (const) reference to the underlying 1-D storage object (can be anything with a std::vector-like API). `data` returns the memory buffer (i.e C pointer to dynamically allocated array) of the expression, while `operator()` performs a strided scheme computation.